pjsip/src/pjsua-lib/pjsua_call.c

```diff
@@ -620,15 +620,23 @@ on_make_call_med_tp_complete(pjsua_call_id call_id,
 
 
 /*
- * Cleanup call setting flag to avoid one time flags, such as
+ * Cleanup call setting to avoid one time flags, such as
  * PJSUA_CALL_UNHOLD, PJSUA_CALL_UPDATE_CONTACT, or
  * PJSUA_CALL_NO_SDP_OFFER, to be sticky (ticket #1793).
+ *
+ * This function will also reset the media direction to default.
  */
-void pjsua_call_cleanup_flag(pjsua_call_setting *opt)
+void pjsua_call_cleanup_setting(pjsua_call_setting *opt)
 {
+    unsigned i;
+
     opt->flag &= ~(PJSUA_CALL_UNHOLD | PJSUA_CALL_UPDATE_CONTACT |
 		   PJSUA_CALL_NO_SDP_OFFER | PJSUA_CALL_REINIT_MEDIA |
 		   PJSUA_CALL_UPDATE_VIA);
+
+    for (i = 0; i < PJSUA_MAX_CALL_MEDIA; i++) {
+    	opt->aud_dir[i] = opt->vid_dir[i] = PJMEDIA_DIR_ENCODING_DECODING;
+    }
```

Comment:
Please make sure if direction needs to be reset by this function, note that this function is to reset temporary/one-off settings/options (rather than setting default values), while media direction is something to maintain over the subsequent in/outgoing re-INVITE/UPDATE, app should manually set the direction setting to change it and should not be surprised by auto-reset after sending/receiving re-INVITE.

Reply:
The specification of the issue is to specify audio/video direction when making/answering call, so I believe it's a one-time setting (will probably need add doc to make this clearer), rather than something that needs to be maintained.

There are some complications and complexities regarding this media direction, especially if it needs to be maintained. This is caused by which index refers to which media?

Note that the settings of audio and video count only specifies **active** audio/video streams. For example: `aud_cnt` is 1 and `aud_dir[0]` is `PJMEDIA_DIR_ENCODING`. Then there are 2 audio SDP, first one inactive, second one active. `aud_dir[0]` will refer to the first active media, i.e. the second media.

But the status and property of the media can change over time. Some of the scenario that can happen:
- the first media activated
- a third audio media added
- the second one becomes inactive after double hold
- remote can change the property of any of the media by adding srtp (note that when answering we sort the media based on the "quality", so the first media is not necessarily the first active media in the SDP).
- and although this is illegal by the RFC, remote can also remove media.

Since it's no longer clear `aud_dir[0]` should refer to which media, that's why I made it a one-time setting and reset it every time. The cons of one time setting is that for every re-INVITE/UPDATE, app will need to specify the media direction again if desired. But I believe this is still much less complicated than trying to understand the index referencing.

Another alternative that I can think of is to make `aud_dir[i]` always refers to the i-th audio media in the SDP. The drawback of this alternative is that there will be a disconnect between `aud_cnt` and `aud_dir`.

Reply:
Agree, I think it needs a specification doc please.

Another idea is to introduce a flag in call setting to specify that directions settings are set and to be applied by the following op. So the `clear_flag()` function will just clear the flag and other op will not reset the media direction.

Regarding media dir indexing, just an idea, perhaps we can just make it one array only for all media types, similar to media info in call info. Yes, a disconnect between `aud_cnt` and `aud_dir`/`med_dir` is okay, just the spec needs to be clear about this, e.g: for making call, audio will take the first part, then video, so med_dir[0] will refer to audio (if any) and so on, for other ops (incoming call, re-INVITE, etc), it seems to be simpler.

Let's discuss this further over Slack if needed.

Reply:
One array for all media is a good idea indeed. It avoids the index complication and is much simpler to implement. So please review the latest change. Thanks.